address-space-controller/src/main/java/io/enmasse/controller/AddressFinalizerController.java

```diff
@@ -0,0 +1,142 @@
+
+    private static final Integer BATCH_SIZE = Integer.getInteger("io.enmasse.controller.AddressFinalizerController.batchSize", 100);
+
+    private MixedOperation<Address, AddressList, DoneableAddress, Resource<Address, DoneableAddress>> addressClient;
+
+    public AddressFinalizerController(NamespacedKubernetesClient controllerClient) {
+        super(FINALIZER_ADDRESSES);
+        this.addressClient = KubernetesHelper.clientForAddress(controllerClient);
+    }
+
+    @Override
+    public Result processFinalizer(final AddressSpace addressSpace) {
+
+        var addressSpaceName = addressSpace.getMetadata().getName();
+        var addresses = this.addressClient.inNamespace(addressSpace.getMetadata().getNamespace());
+
+        log.info("Processing finalizer for {}/{}", addressSpace.getMetadata().getNamespace(), addressSpace.getMetadata().getName());
+
+        AddressList list;
+        String continueValue = null;
+        do {
+            list = addresses.list(BATCH_SIZE, continueValue);
+
+            log.debug("Processing addresses - found: {} -> continue: '{}'", list.getItems().size(), list.getMetadata().getContinue());
+
+            for (final Address address : list.getItems()) {
+                processAddress(addressSpaceName, address);
+            }
+
+            continueValue = list.getMetadata().getContinue();
+            if (continueValue == null || continueValue.isBlank()) {
+                log.debug("Completed cleanup");
+                return Result.completed(addressSpace);
+            }
+
+        } while (true);
+
+    }
+
+    /**
+     * Process a single address.
+     *
+     * @param addresses
+     *
+     * @param address The address to process.
+     */
+    private void processAddress(final String addressSpace, final Address address) {
+
+        if (matchesAddressSpace(addressSpace, address)) {
+            this.addressClient
+                    .inNamespace(address.getMetadata().getNamespace())
+                    .delete(address);
+        }
+
+    }
+
+    private boolean matchesAddressSpace(final String addressSpace, final Address address) {
+
+        return matchesAddressSpaceByAnnotation(addressSpace, address)
+                .or(() -> matchesAddressSpaceByName(addressSpace, address))
+                .orElse(false);
+
+    }
+
+    /**
+     * Check if the address is part of the address space, by matching the annotation.
+     *
+     * @param addressSpace The address space to test for.
+     * @param address The address to test.
+     * @return If the "addressSpace" annotation is missing, {@link Optional#empty()} is being returned.
+     *         Otherwise
+     *         {@code true} will be returned if value of the annotation matches the address space name,
+     *         or {@code false} otherwise.
+     */
+    private Optional<Boolean> matchesAddressSpaceByAnnotation(final String addressSpace, final Address address) {
```

Comment:
I think you can drop this. We make an assumption that address is always prefixed by address space name.